extensions/pkg/controller/worker/genericactuator/actuator_reconcile.go

```diff
@@ -51,11 +51,18 @@ func (a *genericActuator) Reconcile(ctx context.Context, worker *extensionsv1alp
 		return fmt.Errorf("could not instantiate actuator context: %w", err)
 	}
 
-	// If the shoot is hibernated then we want to scale down the machine-controller-manager. However, we want to first allow it to delete
-	// all remaining worker nodes. Hence, we cannot set the replicas=0 here (otherwise it would be offline and not able to delete the nodes).
+	// If the shoot is marked for hibernation then check if MCM needs to be scaled Up/Down. This will be based on the existing number
+	// of machines across all machine deployments for the worker. If there are existing machines that need to be shutdown then MCM should be up.
 	var replicaFunc = func() (int32, error) {
-		if extensionscontroller.IsHibernated(cluster) {
-			return kutil.CurrentReplicaCountForDeployment(ctx, a.client, worker.Namespace, a.mcmName)
+		if extensionscontroller.IsHibernationEnabled(cluster) {
```

Comment:
I rather had the idea to always scale up MCM when the Shoot is hibernating or waking up (`spec.hibernation.enabled` != `status.hibernated`). It should be a more elegant solution than checking the number of existing Machines.

Reply:
My current understanding (please correct it if i am wrong).

`status.hibernated` is not set to true unless and until all machines have been stopped and also the control plane components have been stopped. `actuator_reconciler` is not really responsible for shutting down all shoot control plane components other than scaling down MCM and CA for the shoot cluster marked for hibernation.

Explicitly checking the number of available machines ensures that it does not attempt to scale up MCM when there is another actor (possibly gardenlet) which is trying to bring down control plane components for this shoot after MCM has stopped all machines. Otherwise this would lead to conflicting actions by more than one actor.

Reply:
>The drawback of the current check is that it will always return the current replicas when shoot is hibernating (which can be 0 due to dependency-watchdoing scaling MCM down).

If DWD has already scaled down MCM then it will not return 0. It will first check if there are available machines that needs to be stopped. This will be > 0 if there are machines available. Then the returned replicas will be 1
```go
if numExistingMachines > 0 {
	return 1, nil
}
```
Thanks for pointing me to the helper `controlplane.go` using it i was able to navigate to `gardenlet/controller/shoot/shoot_control_reconcile.go` where `HibernateControlPlane` method is called which will scale down `GRM`, `KCM` and `KAPI` after waiting for nodes to be removed which in turn waits for the machines to be removed.

So to conclude `MCM` and `CA` scale down will only be done by `actuator_reconcile` and no other actor. In that case the above simplified check would be sufficient.
Will make the changes.

Reply:
There is also the following code at the fag end of the `Reconcile` method:

```go
if isHibernated {
   if err := a.scaleMachineControllerManager(ctx, logger, worker, 0); err != nil {
     return err
   }
}
```
The above will scale down MCM and as far as i understand it does not wait for all machines to be shutdown. In the next reconciliation run it will then scale up MCM again.

Reply:
> If DWD has already scaled down MCM then it will not return 0. It will first check if there are available machines that needs to be stopped. This will be > 0 if there are machines available. Then the returned replicas will be 1 and not 0.

Sure, by the "current check" I mean the revision in the master, not in this PR. I think your approach would also work, but if you want let's try to avoid listing Machines and try to implement the proposal from https://github.com/gardener/gardener/pull/6054#discussion_r886340796. As I mentioned above it should be closer to what we do for KCM and GRM.

 > So to conclude MCM and CA scale down will only be done in worker/genericactuator/actuator_reconcile.go and by no other actor.
 
Yes, exactly.

Reply:
sure, i have now changed the approach and removed the need to list machines.